index/encoding/doc/doc.go

```diff
@@ -0,0 +1,274 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Softwarw.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package doc
+
+import (
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/m3db/m3ninx/doc"
+	"github.com/m3db/m3ninx/index/encoding"
+)
+
+// Documents file format metadata.
+const (
+	magicNumber = 0x6D33D0C5
+
+	fileFormatVersion = 1
+
+	headerSize = 0 +
+		4 + // magic number
+		4 + // version
+		0
+
+	trailerSize = 0 +
+		8 + // number of documents
+		4 + // checksum of payload
+		0
+)
+
+const initialEncoderSize = 1024
+
+var (
+	errEmptyDocument      = errors.New("cannot write an empty document")
+	errInvalidMagicNumber = errors.New("encountered invalid magic number")
+	errChecksumMismatch   = errors.New("encountered invalid checksum")
+)
+
+type writer struct {
+	version uint32
+
+	writer io.Writer
+	enc    *encoding.Encoder
+
+	docs     uint64
+	checksum *encoding.Checksum
+}
+
+// NewWriter returns a new Writer for serializing documents to an underlying io.Writer.
+func NewWriter(w io.Writer) Writer {
+	return &writer{
+		version:  fileFormatVersion,
+		writer:   w,
+		enc:      encoding.NewEncoder(initialEncoderSize),
+		checksum: encoding.NewChecksum(),
+	}
+}
+
+func (w *writer) Open() error {
+	w.enc.PutUint32(magicNumber)
+	w.enc.PutUint32(w.version)
+	if err := w.write(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (w *writer) Write(d doc.Document) error {
+	if len(d.Fields) == 0 {
+		return errEmptyDocument
+	}
+
+	w.enc.PutUvarint(uint64(len(d.Fields)))
+
+	for _, f := range d.Fields {
+		w.enc.PutBytes(f.Name)
+		w.enc.PutBytes(f.Value)
+	}
+
+	if err := w.write(); err != nil {
+		return err
+	}
+
+	w.docs++
+	return nil
+}
+
+func (w *writer) Close() error {
+	w.enc.PutUint64(w.docs)
+	if err := w.write(); err != nil {
+		return err
+	}
+
+	w.enc.PutUint32(w.checksum.Get())
+	if err := w.write(); err != nil {
+		return err
+	}
+
```

Comment:
Should this also close the underlying `w.writer`? that way the user can create an `io.Writer` and passing it to the `encoding/doc/Writer` transfers ownership of it

Reply:
At the moment, I'm of the feeling that we should handle the lifecycle of the `io.Writer` outside of the `doc.Writer` in case we want to, for example, encode multiple writers (documents, terms dictionary, postings list) into the same file. I don't feel strongly here though, and I think actually using the writer in practice will help decide here.

Reply:
I don't think that'll happen - more inclined to making a filetype for each type of writer (separate ones for documents v terms dictionary v postings lists).

Reply:
I think it'll be simpler to use if the doc.Writer takes ownership of the io.Writer. That way we only need to track one var's lifecycle instead of two.